### ISSUE TITLE
v. 2.0 release

### ISSUE THREAD

**Issue by KFlash** [cherow/cherow#316]
Awaiting @fkleuver to optimize the code before release of v. 2.0 to NPM.

**Comment by fkleuver** [comment]
We can do minor/patch releases with optimizations afterwards. If 2.0 is done from a functional point of view I think it's good to go and release it to get some feedback as well. Makes it easier to compare perf too.

**Comment by KFlash** [comment]
There are two minor issues left to fix as far as I know. @kflash mentioned it in the bypass and on `Skype` yesterday. There are a few non-failing cases related to (*enum identifier / keyword*) and the eval and arguments (*enum identifier / keyword*) .

It was mentioned that this could be fixed if the token code was changed so the enum (identifier / keyword) was also recognized as an identifier. Same thing for the eval and arguments (*enum identifier / keyword*) . That has to be fixed before a 2.0 release, and `validateBindingIdentifier` updated to deal with the new changes.

Another important fix is the fact that  a parenthezied async  & non-async arrow with and without a block body doesn't communicate with the scope. The connection is broken. 

According to what I understand, this was something that was pending on your optimizations and a solution for the current issue have been described to you on Gitter I think.

Beyond this, we are now reviewing the code and correcting minor errors that we have come across. Maybe by the end of today or tomorrow or review is complete and everything is good to go from our side.

**Comment by KFlash** [comment]
@fkleuver  The code is feature complete, and done from a functional point of view as you mentioned. But there exist a few places that could be optimized. And maybe it should before any release. It depends on how much this parser can parse before it start to kneel down and pray to God to end the misery. We didn't benchmark this, so we don't know. 
But tokens can be merged to avoid too many `if/else` clauses and 
also optimized to avoid frequent branch prediction fails e.g. using a table dispatch in `parseStatement` and `parseStatementListItem` or eventually use a perfect hash in the switch statement instead.
A few tokens we know that could be merged here is the two for escaped keywords. If they are seen as an identifier you wouldn't need the extra `if/else` in e.g. class decl & class expr. And a few other places.

But this is all micro optimization, and is it worth it in the long run? If we choose to go down this path, I also think the entire token code should be changed because V8 on first run would optimize on string values, so there are only a win-win case in benchmarking and not a RL scenario when you do 
`if (token === Token.ForKeyword` instead of `if(tokenValue === 'for'`) unless you use a perfect hash. I can be wrong, so don't rely entirely on my words about this.

**Comment by KFlash** [comment]
@fkleuver We are done for now from our side, so I guess you can take it from here as stated in #303 

> Agree with @axefrog , @KFlash as I told you earlier my plan was to output multiple builds and have rollup do some replacements where particular features are removed from the bundle at build time. I said I would help you fix the performance so don't worry about that. Just focus on spec compliance / tests

I guess this will be done now?

As @KFlash mentioned for you, what's left to fix is the "scope" disconnection so remaining non-failing arrow tests will fail as expected. And so it flatten the expression parsing and performance.

There are still a few non-failing tests, but easier to fix them after your "clean up".

And also documentation is missing, but as far as we understand this should be fixed by the chinese guy @aladdin-add as mentioned in #284 

Please *note* there is a lot of breaking changes in this release and Cherow 2.0 is not backward compatible with Cherow 1.6.x

We are off now, and let you take it from here.

**Comment by KFlash** [comment]
@fkleuver What's the state of this issue and the optimization?

**Comment by KFlash** [comment]
@fkleuver I don't have NPM access to publish a release.

**Comment by fkleuver** [comment]
I can publish it if it's feature-ready, but you were still mentioning some bug fixes that needed to happen right? This would take more time for me than general performance improvements as I'd need to dig into the spec. I cannot put a timestamp on it due to having a lot on my plate for Aurelia right now. We talked about this on Gitter as well.

In any case, just to get everything clear, what is your status going to be for the coming weeks / months? Are you going to be doing any more work in any area or are you fully leaving the thing with me? I don't mind either way, but then I know what (not) to expect. If you are completely stepping out then I will start reserving some time slots starting next week. Otherwise I will wait until the beginning of March and focus on other priorities first.

**Comment by KFlash** [comment]
@kflash isn't around. I am one of his former students. I think he has vacationed indefinitely, and he's celebrating life's happy days in the Netherlands.  So I don't think you can expect him to come back at first. Maybe as a little maintainer sometime in the future. 
He asked me if I could fix the last things he had on his TODO list, and I've done that now. I don't think there is more that I can do considering my absent knowledge of the ECMA specs. I noted above (*it was me who answered you earlier too*) what remains. There are also some non-failing cases left to fix, but this is an open source project and people can contribute to get them fixed if they like this parser.
I'm going to talk to @kflash on Skype this weekend so I can ask him if it's any more important to fix, but I doubt it.
So consider it safe to take over the ship.

**Comment by KFlash** [comment]
@fkleuver This is @KFlash complete identifier code for the "hot path". This never ended up in Cherow.

```js
function scanIdentifierOrKeyword(context: Context): Token {
    let scanFlags = AsciiLookup[currentChar];

    while (index < input.length) {
      let charFlags = AsciiLookup[currentChar];
      if ((charFlags & CharType.WhiteSpace) === CharType.WhiteSpace) break;
      scanFlags |= currentChar > 127 ? CharType.SlowPath : charFlags;
      advance();
    }

    if ((scanFlags & CharType.SlowPath) !== CharType.SlowPath) {
      value = input.slice(start, index);
      if ((scanFlags & CharType.NoKeywordCandidate) === CharType.NoKeywordCandidate) {
        return Token.Identifier;
      }
      const len = value.length;
      if (len >= 2 && len <= 11) {
        const keyword: Token | undefined = descKeywordTable[value];
        if (keyword !== undefined) return keyword;
      }
      return Token.Identifier;
    }
    return scanIdentifierOrKeywordSlowPath(context);
  }

  function scanIdentifierOrKeywordSlowPath(_: Context): Token {
    return Token.Identifier;
  }
```

**Comment by fkleuver** [comment]
I see what he's trying to do there, but I'm fairly sure there is no performance gain in using `scanFlags` like this. Sure, it makes the code base more consistent (perhaps to some eyes even more elegant), but if raw performance is the goal then a few raw comparisons and boolean variables will likely perform better across all browsers.

Apart from that, I've always said that the scanner code is the wrong thing to be focusing on at this point. It is the easiest thing to look at and try to optimize, but the real bottlenecks are object creation, polymorphism and branching in the parser code.

- Object creation is hard to optimize, but some gains could be made by using a few tactically chosen classes instead of inline objects. That's a trade-off between perf and lib size, though.

- Polymorphism as a bottleneck in cherow mostly boils down to ensuring things that are supposed to be SMI's, are truly always SMI's and not oddballs or worse. I believe there are a few hot paths where both numbers and strings pass through. Not sure though, haven't checked the latest code thoroughly yet.

- Branching is the thing he has been focusing on a lot and the main thing he asked me to look into and optimize. Optimizing this also includes figuring out where unnecessary checks are happening, and removing those. This, in turn, might reveal opportunities for reordering logic in a more optimal way which, if the tests are not 100% accurate and reliable, could wreak havoc on the long term if it breaks something unknowingly. This is why I'm a little reluctant to start working on this right now (besides me just being busy) because if it does break something, even more work will come from that. It will spiral out of control with a parser like this.

So, that's why I prefer to confirm first that cherow is indeed fully spec compliant to the best of everyone's knowledge, according to tests, and the 262 suite, and then release 2.0 so the "pressure" is off. Then I can take my time on the long run to go through everything, add tests where needed, and so forth.

Regarding the things that are still not working, are you able to fix them or what should I expect there?

**Comment by KFlash** [comment]
Hi. I fully understand your point of view, and I hope you understand I'm just a drop-in maintainer and doesn't have a very in-depth knowledge of the more technical aspects of Cherow.  I know what works and what doesn't. And there are no critical things that will be destroyed if you start working on this. What remains is dependent on your optimization according to what I understand on @kflash

For arrows and async arrows the "scope" is broken. Soon as it is fixed, 50 non-failing tests will fail as expected. Then you have 60 tests left to fix.

This tests are e.g. yield doesn't fail every places it should etc. Actually an easy fix with bitmask fiddling.
So first thing to focus on is to get the scope working, and after what @kflash told me the plan was to "flatten" the entire expression parsing into one function - something you should do?

Then you just merge array lit and array pattern into one function and same with obj lit and obj pattern. So you have something

So then there would be something like this

```js
function parseObjLitPattern(state, context, scope, isPattern) {
if (isPattern) return pattern;
return expr;
}
```
The 'scope' today - if no scope exist - is represented by a negative number. `-1`. 

So after what I understand - if you flatten the entire thing you could do something like this I guess

```js
function exprParsing(state, context, scope, origin, isPattern) {
case '{':  `parseObjLitPattern`
}
```
And to get this working in arrows is then easy - you already have the scope.
`exprParsing(state, context, scope, origin, false) `

And this should be fine. But I have no further knowledge of this, nor do I know how you to plan this. This was just an example code from me. But I think you get the point?

Except for this is a few non-failing arrow tests exist, but I think if you flatten everything they can be passed as a boolean instead of the mutual parser flag or you use another set of bitmasks. If that is set up correctly I can get them to fail if you wish.

This is what I know remains to be fixed and optimized. And nothing I can do. I'm sorry for that.

**Comment by KFlash** [comment]
@fkleuver I also forgot to mention that there are a few non-failing enum keyword tests and eval and arguments. According to what I understand, this is related to tokens and not something I can fix. In the source code, these things are marked with "TODO (fkleuver)" so it is easy to see where they should fail.

**Comment by KFlash** [comment]
@fkleuver I looked through the source code now and I can't see other things that need to be *fixed*. There exist around 110 non-failing Test262 suit tests, but they will fail if the code get changed in the way I mentioned earlier.

However there can  be a few regular expression edge cases I may have overlooked - the code doesn't parse regular expression as expected. If you encounter an issue like that, you need to set the `| Context.AllowPossibleRegexp` bit before you trigger the scanner and the issue is solved.

In his original code @KFlash also uses `Context.AllowTemplate` and parses the template way different so it has been a little confusing for me.

Except all this I think there is no more.

**Comment by KFlash** [comment]
@fkleuver If it's okay for you I add - when I find it useful - minor tweaks to the code that I have seen in the experimental code. Like what I did for the lower / upper case letters tweak.

You are starting on this next week as mentioned above?

> If you are completely stepping out then I will start reserving some time slots starting next week.

**Comment by fkleuver** [comment]
I can start working on optimizations yes, but I'm not sure if I'll get around to fix the issues quite yet.

**Comment by KFlash** [comment]
I guess it will take awhile to fully understand the code before you can fix the issues?

**Comment by KFlash** [comment]
@fkleuver I'm not sure how useful it is but I added a way to extend the lexer as you can see in https://github.com/cherow/cherow/commit/1315e96ca9eca92b5780da0e35e28f71e57cc9bb

I'm not an performance expert, and didn't test it. You may improve it, but I think this is one way to extend the parser as I noticed was mentioned in #292.

Once again I "*copied*" this code from @KFlash private code, and I have seen that he can use different callback that returns different results during a simple parse

```js
const lexer1 = function() {}
const lexer2 = function() {}

nextToken(state, context, lexer1);

nextToken(state, context, lexer2);

```

**Comment by fkleuver** [comment]
No, it just requires a big uninterrupted time slot to focus on the code and I don't have many of those at the moment. Performance improvements lend themselves for a few hours here and there, fixing parser bugs takes a full day or two. You know how it goes

**Comment by KFlash** [comment]
I may think improving the error tracking / messages also would be a good thing.  In v. 2.0 @KFlash added an "*arrow*" to pinpoint exactly where the error is, but not always correct.

In his experimental code I noticed (*for edge cases*) he pass the index, line and column values directly to the error func to get exact error loc. 

Like this:

```js
     SyntaxError: Multiline comment was not closed properly (1:3)
/**
   ^
```

**Comment by KFlash** [comment]
@fkleuver I doubt there are more that I can do here, so I let you handle everything from now. I tried to "*adopt*" some code from his experimental code to Cherow, but not that easy beacuse it's two different worlds. He uses a combination of functions and table lookup inside the function bodies. 
So I wish you good luck!

**Comment by fkleuver** [comment]
Speaking of that particular example, we need experimental support for parsing decorators. A lot of people ask about this.

**Comment by KFlash** [comment]
How to implement it?

I noticed this code in his experimental code, and this may be a solution for the "scope" issue I mentioned to you before - if not flatten everything into one function?

```js
function parseArrayLiteral(context: Context, scope: ScopeState, isPattern: boolean) {
    token = scan(context);
    const elements: (ESTree.SpreadElement | ESTree.Expression | null)[] = [];
    while (token.token !== Token.RightBracket) {
      if (token.token & Token.IsIdentifier) {
        elements.push(parseMemberExpression(context, parseIdentifier(context)));
      } else if (token.token === Token.LeftBrace) {
        elements.push(parseMemberExpression(context, parseObjectLiteral(context, scope)));
      } else if (token.token === Token.LeftBracket) {
        elements.push(parseMemberExpression(context, parseArrayLiteral(context, scope)));
      } else if (token.token === Token.Ellipsis) {
      }
    }

    return isPattern
      ? {
          type: 'ArrayExpression',
          elements
        }
      : {
          type: 'ArrayExpression',
          elements
        };
  }
```

**Comment by fkleuver** [comment]
We had it in 1.6.x somewhere. It's relatively simple - as far as syntax is concerned, a decorator is just a `CallExpression` or `Identifier` (that's the decorator syntax without the parens) preceded by an `@` symbol.

It can be placed before a `ClassDeclaration`, or properties / methods / parameters. The decorator itself is then simply added to a `decorators` array of the AST node in question.

TypeScript took the "simple" aspect a step further and added an optional `decorators` property to the base node, so almost everything has a `decorators?: ...` property. Forward compatibility wise that could be a smart move until the spec has finally settled, since it's still not certain what they can really be placed on when the spec is final.

**Comment by fkleuver** [comment]
@KFlash I have done some experimenting with that "everything in a single function" idea in the Aurelia repository but it turns out it's actually a performance killer in V8.

It seems storing stuff on properties of a class can greatly outperform (by a factor of 2-4) loose variables declared in function scope if the function is really big and has many variables, but this applies mostly to isomorphic properties. This has to do with the way V8 optimizes generated hidden classes.

Many small functions does indeed seem to perform better, in particular when the parameter types are isomorphic as well.

**Comment by KFlash** [comment]
@fkleuver I had a Skype chat with @KFlash 

> Here is my findings yes after experimenting

>- lexer code in function scope is fast and can perform good, but only allow you to use
>  a table lookup inside a function scope 2nd level deep or deeper (nested). Hard to maintain.
>
>- Table lookup is fast, but it take time on "first run" to create the needed tables. Smaller tables
>  2nd level deep for one-char punctators and ASCII lookup only and a small switch statement for the 
>  rest is the best choice, See V8 and other engines for comparison.
>
>Also avoid to directly return values. E.g. for a "nextChar" function it's faster if you do 
>
```js
function nextChar(state: ParserState) {
  state.currentChar = 0;
  state.column++;
  state.index++;
  if (state.index <= state.length) state.currentChar = input.charCodeAt(index);
}

>// And then
>
if (state.currentChar) ...
```
>This has been tested by Chakra team among others, and it has been done multiple benchmarks on it.
>Also V8 itself do similiar things.

>Yes, I managed to come up with a lexer code that outperform Cherow 2.0. On 250 lexer tests my code
>uses average of 70ms. Cherow 2.0 uses 200ms for the same.

>For the "issue" regarding scoping. Is that an issue? Just move the scope to the scanner state and it's
>available everywhere you need it instead of passing as a function arg. I think @fkleuver knows how to >fix that.
>Than man is a genious :)

**Comment by KFlash** [comment]
@fkleuver It's me. The real one :) I'm just temporary here to clarify a few things.

I noticed you mentioned hidden classes.  In my research and experiments I found out that if we keep the parser state as a pure object literal, we gain performance with 3% vs a class alternative. You can try it yourself in various benchmarks, and you will notice the performance will drop.

Regarding the function scope thing. If you want the code to be properly scalable, and also care about memory usage a class is the best choice, **but** as in TypeScript case they have to create a new class instance every time you do a change in the code, and then do a lot of comparison. That is expensive. Use of the `new` keyword can be a performance bottleneck. In Cherow I can't see any disadvantage of using classes a few places in the code, and it will help reduce memory usage. So you are right there.

Another thing we should be aware of is to avoid using double-precision floating point number, and instead use 31-bit signed integer, or bit masks. This because of how V8 internally treats numbers. 
And if things are only locale within one function - use a local boolean instead. This is V8 related, and I'm not sure how the real performance differences are in other JS engines, but V8 is the most common one and also used by NodeJS so we should focusing on optimize for V8.

And we should prefer monomorphic over  polymorphic!

Soon as I get to a computer I will try to push to a new branch my private lexer code so you can see how I do it when I optimize things for real. Still that code is "*unpolished*" and need improvements, but I changed a few things from the original `Cherow` code. Example the error handling.

You also mentioned decorators. It's still on stage 2, and the specs keep changing. Therefore I was hesitant to implement it. And *if* this should be implemented it has to be under the experimental option. But does V8 support decorators? I haven't checked it recently.

The experimental option should be equal to V8's experimental feature flag.

Anyway. It's an easy thing to implement, and I'm sure you can do it if you find time for it :)

The last thing to mention is the non-failing cases. When I left it was around 130 tests that didn't fail as it should, and I know it's down to 110 now. J.K can fix this things. I'm sure. But before that we need to solve the scoping issue.

The scoping issue for arrows can be solved if you

1) Move the scope to `ParserState` (*be aware of performance*)
2) Try to do something similar that I noticed `J.K` pasted from my experimental code

I'm offline again now, but I'm 100% confident that Cherow is in safe hands with you !

**Comment by fkleuver** [comment]
@KFlash The difference between plain objects and classes is not particularly large unless there are methods involved, but a class does use a little bit less memory per instance. Where is that benchmark showing the perf drop?

Regarding 31-bit signed integer, yes, and perhaps we can use `Int32Array` instead of `Uint32Array` for the various lookups.

Regarding differences between engines, Chakra and Gecko benefit a lot more from directly using natives when possible (e.g. `TypedArray` over a normal array with only numbers) since V8 tends to optimize things in that direction anyway, where other engines don't. I noticed this when I tried to optimize a lis algorithm (there are more factors in play, but TypedArray hardly made a difference in Chrome whereas it almost doubles/triples perf in Edge and IE11: http://jsben.ch/MoAdv)

The only thing that will change about the decorators spec and how that would affect the parsing logic is where they are valid/invalid. So, if for starters we don't attempt to validate them (wait with that until the spec stabilized) then we won't need to keep running after the spec and in the meantime folks can parse decorators.

It's not implemented in any runtime, but folks want to parse code with decorators (which works after transpiling with ts/babel), and cherow fails to. That's a reason for many people to opt for ts or babel instead of cherow, when really they would prefer cherow for performance and precision considerations.

**Comment by KFlash** [comment]
There are no specific benchmark. I just ran the same benchmark that exist on the demo page. With and without "cold" JIT. It always perform worse when I was using classes.

Maybe an good idea to use `Int32Array` for lookup, but in my experimental code I moved away from the large table lookup and are using multiple smaller table lookup. E.g. 2 different tables for number and identifiers and it did give a performance boost. And similar to how V8 does it.

```js
 // `a`...`z`, `A`...`Z`, `_`, `$`
  if ((AsciiLookup[state.currentChar] & CharType.Letters) > 0) return scanIdentifierOrKeyword(state, context);

  // `0`...`9`
  if ((AsciiLookup[state.currentChar] & CharType.Decimal) > 0) return scanNumericLiterals(state, context, false);
```
I did stop experimenting with this a while ago because I focused on  finalizing Cherow 2.0. And I'm using a similar lookup for one char token for punctuators.

Regarding decorators. Recently I also noticed they changed the name on some hooks etc, but I guess that hasn't any impact on parsing. So yes I agree that we should implement it and hide it under the experimental option. It's still Stage 2, so not qualified *yet* for the `next` option.

**Comment by KFlash** [comment]
@KFlash I added my private experimental lexer code. [Can be found here](https://github.com/cherow/cherow/tree/experiment/src). It's just an experiment, and even if it works don't expect miracles with it :) The only thing that doesn't work is `identifier + surrogate` and `surrogate + identifier`. Just me being lazy :)

**Note:**  This was the code I experimented with before what become Cherow 2.0, and the perf are actually improved over current 2.0.

**Comment by KFlash** [comment]
@fkleuver I updated my private experimental code to latest. Use it if you want when optimizing Cherow 2.0. Most of the "hot path" have been optimized and now uses a table lookup. The performance is good, but ofc. this can be optimized further if you chose to use it. Example numbers in the ASCIILookup table can be separated so you can do a fast lookup for octals and binary as well to mention a few. But this is all micro-optimizations on a very low level and are kind of pointless.  I rather focusing on optimization for significant digits and SMI since that is what is most used. Same for identifiers. I improved the "hot path" for "*plain*" identifiers, but not for escapes. And so on.

Anyway. My state from now on is not to be available or online. J.K will use my profile from time to time and just ping him if you want him to fix things for you.

**Comment by KFlash** [comment]
@fkleuver When can we expect a v. 2.0 release?

**Comment by fkleuver** [comment]
Not sure yet, I'm first going to make benchmarking a bit easier/quicker to do (with more reliable results) so that perf tweaks don't take so long to validate.

**Comment by KFlash** [comment]
@fkleuver Sounds interesting! Btw. What have been suggested before regarding solution for getting the scope working in arrows - I only think that can be fixed if you move the scope to the state object. Not sure about the performance regarding that. 
Merging obj lit and obj pattern into one will give you too much headache, and not worth it. If it's possible at all. And if you manage to fix it somehow, what's the benefits? You still can't avoid to re-interpret to pattern (*wich can be expensive*),  but you can use it to reduce re-interpretation if you example do `type: isPattern ? 'ObjectPattern' :  'ObjectExpression`.

Another thing you could do is to optimize the cover grammar implementation and avoid the function calls if possible. And also optimize the scope implementation.

And I say you should stick to current parser flow and not try to flatten anything or merge anything into same function. The parser flow Cherow currently are using,  is well tested by many parsers as well as V8. Doing it differently will just cause more headache, and what I tried to show you from my experimental code regarding expression parsing. Well. I may suggest not to walk down that path. Too many wormholes to get out of :) Plus you would need lot more bit masks to keep track of the different states. A simple boolean will not help you :) And this things will turn out to be too complex for open source.

I have experimented with a lot of things and in the end you will find out that the ESTree specs is what will bite your ass. It's so bad designed in many ways :(

**Comment by KFlash** [comment]
@fkleuver Regarding the wormholes I mentioned above. There is no performance benefits to try to refactor Cherow into something like that, and it's *nearly impossible*. I guess that why V8 and others didn't write their parsers that way. It's actually to hard :(

Anyway. Just to guide you. The benchmark results below is Cherow 2.0 *vs* that wormhole idea I just mentioned

![array_bench](https://user-images.githubusercontent.com/31855118/52892128-c3d23980-31ca-11e9-8a2a-bc9882bc4f54.png)

![obj_bench](https://user-images.githubusercontent.com/31855118/52892139-e1070800-31ca-11e9-86a7-19029ccd4cce.png)

You see? There is no performance benefits doing it.

**Comment by KFlash** [comment]
@fkleuver What do you think about something like this? This is taken from my "too many womholes and  headache experiment"

```js
/**
 * Parse array expression or pattern
 *
 * @param {ParserState} state
 * @param {Context} context
 * @param {boolean} isPattern
 * @returns
 */
export function parseArrayExpressionOrPattern(
  state: ParserState,
  context: Context,
  isPattern: boolean
): ESTree.ArrayExpression | ESTree.ArrayPattern {
  nextToken(state, context);
  const elements: (ESTree.Identifier | ESTree.AssignmentExpression | null)[] = [];
  while (state.token !== Token.RightBracket) {
    if (consumeOpt(state, context, Token.Comma)) {
      elements.push(null);
      if ((state.token as Token) === Token.RightBracket) {
        break;
      }
    } else {
      let expr: any;

      if ( (state.token & Token.IsIdentifier) === Token.IsIdentifier) {
        expr = parsePrimaryExpressionExtended(state, context, false);

        if (consumeOpt(state, context | Context.AllowRegExp, Token.Assign)) {
          if (!state.assignable) report(state, Errors.UnexpectedToken, 'Donna G. F');
          expr = {
            type: 'AssignmentExpression',
            expr,
            operator: '=',
            right: parseExpression(state, context)
          };
        }
      } else if (state.token === Token.LeftBrace) {
        expr = parseObjectExpressionOrPattern(state, context, isPattern);
      } else if (state.token === Token.LeftBracket) {
        expr = parseArrayExpressionOrPattern(state, context, isPattern);
      } else if ((state.token & Token.IsStringOrNumber) === Token.IsStringOrNumber) {
        expr = parseLiteral(state, context);
      }

      elements.push(parseAssignmentExpression(state, context, parseMemberExpression(state, context, expr, false)));
      if (consumeOpt(state, context, Token.Comma)) {
        if ((state.token as Token) === Token.RightBracket) {
          break;
        }
      } else {
        break;
      }
    }
  }
  consume(state, context | Context.AllowRegExp, Token.RightBracket);

  return {
    type: isPattern ? 'ArrayPattern' : 'ArrayExpression',
    elements
  };
}
```

**Comment by fkleuver** [comment]
> The parser flow Cherow currently are using, is well tested by many parsers as well as V8

That doesn't sound like a you thing to say. Isn't the point of cherow to rely on such extensive testing that it doesn't *need* to follow well-known patterns, and can instead attempt to do better?

> **But** this is doomed to fail at some point

Not if properly tested. I was trying out [stryker](https://stryker-mutator.io/) to find out exactly which logic in the code base is redundant, and use that information to try and optimize things further.

Now my problem with your hesitation is that, if cherow was indeed sufficiently tested (and you were truly confident about that), you wouldn't be hesitating. If even you are hesitant, then what am I supposed to be, knowing less about the spec?

So let's think first of a strategy on how to be sure we can really rely on the tests then. We need to be able to.

**Comment by KFlash** [comment]
I'm not hesitating regarding the existing tests. They cover everything in Cherow and there are more than enough tests to be confident that the parser works as it should. No problems there :)
I'm only hesitating about my experimental code because I know that code will fail somehow or I will run into edge cases that I can't solve (*already experienced it*). That's the reason why I let Cherow follow the well-known pattern. That way we will always know the flow and can optimize from there, and we will also always know that the AST are build properly. Example `LeftHandSide + member expr + primary expr + member expression` is always a pitfall. In Cherow we will always know that this will solve cases like `a().b, b[].c.d.e` without any issues, And it will always happen. 

In my experimental code this has to be done manually. What I take for granted in Cherow, turns out to be a huge failure sometimes in my experiment.  The main reason for my hesitation.

So when it comes to the tests.I trust them, and even if there should be a missing tests, we still know that the code will work because the pattern we follow is set up to work correctly. So it's actually enough with 2 - 3 tests to know that member expr works as expected, and if it does. Well. Then also 1200 other tests that rely on member expr will work.

**Comment by KFlash** [comment]
@fkleuver Just now I ran into an issue with my experimental code while trying to parse out
 `[x, ...z + arr, y]`.  The cause of this issue is the missing binary expression.
With Cherow 1.6x and now Cherow 2.0 this kind of issues will never happen because we are following the well-known pattern, but in my experiment this has to be done manually. And how many cases like this I haven't discoverd **yet**? 

So no worries about Cherow 2.0 and the tests. They cover what it needs and the actual test case above was copied from Cherow own tests.

```js
+ expected - actual

             "expression": {
               "elements": [
                 {
                   "argument": {
      -              "name": "z"
      -              "type": "Identifier"
      +              "left": {
      +                "name": "z"
      +                "type": "Identifier"
      +              }
      +              "operator": "+"
      +              "right": {
      +                "name": "arr"
      +                "type": "Identifier"
      +              }
      +              "type": "BinaryExpression"
                   }
                   "type": "SpreadElement"
                 }
               ]
```

**Comment by fkleuver** [comment]
I don't understand. If you ran into an issue that the tests didn't expose, then doesn't that mean there are still missing tests?

**Comment by KFlash** [comment]
@fkleuver I think you are confused with my experiment example? Just think Cherow. The well-known pattern goes like this

`assignment expr -> member expr -> primary expr -> member expr (*again)`

The [MemberExpression](https://tc39.github.io/ecma262/#prod-MemberExpression) are quite limited. 

So if you parse out something like `([a.b.])`, the "*tail*" is there. The `.b`. But let us say you are missing a test for this case. Still no reason to get panic.

Let say you have a test case for this one `foo().a.b.c`. You actually testing both programs and have a test that cover them both. 

This is "freebies" when you follow the well-known pattern. You can always predict, and what pass for one case has to pass for the other one too even if there are no specific tests for it.

Got it?

In my experimental Kleuver code I moved away from this pattern and experimenting and ended up in the situation I illustrated earlier.

**Comment by KFlash** [comment]
@fkleuver One thing that actually could improve performance is a refactor of `parseAssignmentExpression` or at least clean it up, but note that Async arrow is a PITA to parse.  An ideal solution would be something you see in V8 [here](https://github.com/v8/v8/blob/master/src/parsing/parser-base.h#L3002) and [here](https://github.com/v8/v8/blob/master/src/parsing/parser-base.h#L1657) and the specs for it [here](https://tc39.github.io/ecma262/#prod-CoverCallExpressionAndAsyncArrowHead)
But because of the way ESTree have been designed and the fact that non-async arrow doesn't like the call expression, but the async arrow does. Well. You got one problem. And then if you try to nest them or have a serie of arrows after each other (*there exist tests for it*) you get another problem. And so on... I tried to come up with a OK solution based on the limitations I had when implemented this. 
In Cherow 1.6.x we have a different solution, but that solution have flaws. There are a few cases that doesn't parse at all as it should :) I discovered it while fixing this in Cherow 2.0.

A world without a call expression in this cases would have been the best :)

**Comment by fkleuver** [comment]
One thing I noticed in running the benchmark is that both TypeScript and Aurelia are particularly slow to parse in comparison to 1.6. What those 2 files have in common (which the other files don't) is very large functions (in TypeScript they are simply large functions, in Aurelia they are `iife`s). I've got the feeling that this slows cherow 2.0 down so much due to the scope tracking. But I'm not sure why this would be so much slower in function scope than it would be in module scope. It's really a huge difference, up to 50% extra perf loss

**Comment by KFlash** [comment]
When you are parsing a function in Cherow 2.0 you create 3 scopes, and append 2 scopes. You only create 2 scopes if you are in function expression. You validate the function name in the scope too. In the formal list it also keep track if it is a complex arg or not. Not sure if this have the huge impact?
Did you try to profile it?

**Comment by KFlash** [comment]
@fkleuver Did you try to do a comparison vs Acorn? What's the differences? 
Try to either disable or improve the scoping, maybe it help?

I can try to experiment with a different scoping implementation soon as I get to a computer, but in Cherow 2.0 you may move it to the state object so it will work for arrows and async arrows. No way around that because of the well-known pattern we follow, and you can't pass it as func args all the way down to primary expr.

But try lazy loading?

**Comment by KFlash** [comment]
@fkleuver The function args are pattern, so I may think if you use the `inArgList` bit mask, you can avoid to reinterpret to pattern in `AssignmentExpression` everytime you parse out a function. Maybe that help?

**Comment by KFlash** [comment]
@fkleuver you will add decorators before releasing?

**Comment by KFlash** [comment]
@fkleuver Can you try to think about a better algorithm for this one? E.g. escaped keywords.
See [V8](https://github.com/v8/v8/blob/master/src/parsing/parser-base.h#L2923). In a few places we still need to fix this, so something like this?

```js
const awaitPos = state.startIndex;
  let expr = parseIdentifier(state, context);
  if (awaitPos + 5 !== state.startIndex) report(state, Errors.UnexpectedToken, 'escaped');
```

I'm not on a computer, so I can't test, but can you try to come up with something better? I doubt relying on the token length is a good idea. It's different in Cherow.


And you add decorators?

**Comment by fkleuver** [comment]
Did some profiling in V8, seems to be spending a lot of time in `recordTokenValue` if I'm not mistaken. It seems to be a bit tricky to get a good overview in NodeJS directly though.

[output-all.log](https://github.com/cherow/cherow/files/2882803/output-all.log)

Yeah I don't know if relying on token length is a good idea at all, because what about escape characters? Or what was this about?

**Comment by KFlash** [comment]
`recordTokenValue` is the main function for scoping. Records everything including lexicals and get called when using lexicals, functions, arrows etc. So I am not surprised about that. Ideas how to refactor & optimize it?

An idea for scoping is just convert it to class, and rely entirely on bit masks. 

Regarding escaped characters. Almost everything related to it is fixed. There exist a few edge cases like for `await` that doesn¨t fail. So I was thinking if doing something like mentioned earlier would fix the non-failing  tests. I'm not on a computer now so can't test it but you see how I did check for directive prologues? Take a look at `parseStatementList`.

**Comment by KFlash** [comment]
@fkleuver Do you think this is an faster way of implementing the grammar? I found some time to work on my experiment, and still not sure this will work...

```js
 if ((state.token & Token.IsIdentifier) === Token.IsIdentifier) {
      expr = parsePrimaryExpressionExtended(state, context, false);
      if (consumeOpt(state, context | Context.AllowRegExp, Token.Assign)) {
        const assignBefore = state.assignable;
        state.assignable = Assignment.Assignable;
        const right = parseExpression(state, context);
        let assignable = (state.assignable | Destructuring.Cant) ^ (Destructuring.Cant | Destructuring.Assign);
        state.assignable =
          (assignBefore |
            ((assignable | Assignment.NotAssignable | Assignment.Assignable) ^
              (Assignment.NotAssignable | Assignment.Assignable)) |
            Assignment.Assignable |
            Assignment.NotAssignable) ^
          Assignment.Assignable;

        expr = {
          type: 'AssignmentExpression',
          left: expr,
          operator: '=',
          right
        };
      } else if (state.token === Token.Comma || state.token === Token.RightParen) {
        if ((state.assignable & Assignment.NotAssignable) === Assignment.NotAssignable) {
          state.destructable |= Destructuring.Cant;
        }
        state.destructable |= state.assignable;
      } else {
        state.destructable |= state.assignable;
      }
```

**Comment by KFlash** [comment]
@fkleuver What's your progress with v. 2.0? A release soon? My suggestion / idea above isn't possible because of the way Cherow is designed, and I doubt it's faster. Not even V8 handle the grammar like that. Anyway. I got it working like that in my experiment for array and rest & spread, but doubt it's faster. And a lot more code is needed to fix it :(

**Comment by KFlash** [comment]
@fkleuver Ideas on how to improve the scoping so you can release v. 2.0 soon?

**Comment by fkleuver** [comment]
@KFlash After giving it some thought I came to the conclusion that scoping would be better done *after* parsing as a separate, optional step, by walking the parsed AST and analyzing on that instead of directly on the tokens during parsing. There are a few reasons for that:
1. The obvious one: performance. If you disable scoping, 2.0 might very well even be faster than 1.6
2. Lib size when not using scoping: if you don't use the feature, the scoping logic can be tree-shaken. Alternatively, we could distribute scoping as a standalone importable file so one would not need to rely on bundlers to do this
3. Improved API design and test coverage from an integration perspective - by having cherow "dogfood" the AST output that cherow itself produces in order to perform post-processing, we might run into potential issues/limitations and are able to fix those before exposing end users to them (or we might conclude that everything is good and be more confident to release)

**Comment by KFlash** [comment]
@fkleuver Interesting idea! I guess you are thinking of using a pure binary search to walk the AST? It might be faster that way, and just swap the state on the flags as you go and throw / not throw in the end based on the bit mask state. Something like 

```js
let const = 1;
mask |= (let | const) // invalid

// In the end throw errors based on masks state
switch (mask) {
  case 'let':
         if (mask & const) // throw
}
```
I did similar kind of thing in my experiment when I implemented the grammar, but I had to "play" with arrows, arrays, objects and parens at the same time and if the state changed for one of them, it would give different result. Plus I use 3 different masks to keep track of the state.

It was a PITA to implement, but for scoping it's easier because you walk the AST and mark it as you go. However you would need a 2nd set of masks I guess to figure out if you are in global, block or function body / head. With the extra masks you could easy also track the stupid lexical vars behavior. 

Once again. It's doable and sounds like the best way to go. It would also solve the arrow + scope broken thing :)
And to let the end-user choose if he / she want to use the scoping is the right thing to do. V8 already have scoping implemented, and some may prefer to use the native one, and others just need the performance Cherow give without scoping.

How long time would you use to implement the this change?

**Comment by KFlash** [comment]
If you implement this a natural new breaking change that could land in 2.1 would be to remove the AST itself from the Cherow core and instead use binary values for the AST. Then in the end when you walk the AST, you simply assemble it again based on the state of the binary masks and do lexical scoping as well. 

1. You would be able to output different AST's. Not only ESTree
2. Lexical analyzes can be done without any parsing (*and still solve the regexp issue*) 
3) Use Cherow as a validator only. Analyze the input without parsing and if no errors, well everything is okay. That way Cherow would be super fast :)

**Comment by fkleuver** [comment]
You want to define a binary notation (something like ProtoBuf) to store the parse result in an ArrayBuffer or the likes? So `start` and `end` position + NodeType. And then parse *that* back into an AST upon request. So if you only need validation, no AST is produced. Lol. I like the idea but I think most people do actually want the AST, and this would put an extra parsing step in between.
It is a really fun idea though, I'll give you that.

**Comment by KFlash** [comment]
I didn't think it through, and as you know I'm experimenting a lot. I was more or less thinking like this
`return statement | binary | member | primary | identifier /* additional masks for assignment & destruct */ | destructible`

Then I guess you could do a similar lookup as done for tokens & errors? Where the main AST nodes are binary masks & sub masks and the rest is simple enums.

I never thought it through, but I may have a huge Cherow change for v. 2.1, but maybe not this. It all depends if I get done with experimenting, but I have already done the impossible things with my experiment, and I *broke free* from the default standards and still producing the same ESTree output as Cherow. Not even V8 is close to what I actually have done so far :)

**Comment by fkleuver** [comment]
I think the main point of improvement at the moment is to cut loose the scope validation in the first place, as per 80/20 (or in this case more like 95/5) rule. Almost half of the time is spent doing that. It's huge.
If scope validating winds up being a little inefficient compared to what it is now, I don't think that's a big deal as the performance would still roughly be doubled for the majority cases.

**Comment by KFlash** [comment]
Agree that the main focus has to be to cut loose the scope validation, and back then when I implemented it I was focused on getting things working as you suggested, and performance after. The `any` type is either a number and object. Something I did in a rush. The main purpose is that for expression parsing it should do nothing with the main scope. Only add names and lexicals.  The plan was to use a second arg or a scope masks names `NotScopeAble` and then `if (x === NotScopeAble) return; // does nothing`. But I never added the extra arg :)

**Comment by KFlash** [comment]
@fkleuver As I said I may have huge change for Cherow 2.x, but I'm not in open source mode anymore so confused. Is this readable / understandable for the end-users?

```js
export const enum DestructuringState {
  None = 0,
  CanDestruct = 1 << 0,
  Required = 1 << 1,
  // Only destructible if assignable
  Assignable = 1 << 2,
  // __proto__ is a special case and only valid
  // to parse if destructible
  SeenProto = 1 << 3
}

export const enum AssignmentState {
  None = 0,
  Assignable = 1 << 0,
  NotAssignable = 1 << 1
}
```

**Comment by KFlash** [comment]
@fkleuver You know. In the moment everything become a bit mask, you have lost contact with reality and no longer a human, but a computer yourself :)  
Here you see the benchmark on my non-optimized experiment. Mostly everything is bit masks.

Comparison vs. Cherow 2.0

![kleuver](https://user-images.githubusercontent.com/31855118/53392565-bdb33880-39d4-11e9-9167-48629c3f1cd4.png)

**Comment by KFlash** [comment]
@fkleuver Any progress on perf improvement and the scoping? That has to be the main focus before 2.0 release. Don't worry too much about the few non-failing cases that still exist. I will make sure they will be fixed in v. 2.x sometimes in the future. It depends ...

As you know I have been experimenting, and I think I made the impossible possible. And I plan to release my experimental code to the public one day (*not done yet*). 

My experiment was successful! What I actually did was to parse backward. In that way I have full control on what I should parse, and what should not parse. And I can choose to use *only* the pieces of the parser code I need. Example let say we parse `1 = 2`.

The parser code will do this:

```js
1. Parse out number (1) and set the parser to not assignable
2. Throws an error
3. ... nothing
```
Result:

```js
Unexpected token '=' (1:2)
1 = 2
  ^
```
**Note:** In current Cherow we actually parse out everything before validation and throw an error! We miss this kind of control.

There are a lot more into this, and I still need to play more with this code before I can go public with it. And also the complexity of the bit masks I'm using for the grammar, destructuring and assignments need a re-think before I can release this to open source. 

The entire parser have become a bit mask! And I think maybe this is hard to understand?

```js
assignable =
              state.assignable |
                  ((assignable | 2 | 1) ^
                      (2 | 1));
```

Anyway. The non-failing tests in Cherow 2.0 is already fixed in my current experimental code. So relax on that one!

**Comment by KFlash** [comment]
@fkleuver I experimented a little, and as you said. We get a huge performance boost if you cut loose the scope validation. But with location tracking enabled the performance seems almost identical to v. 1.x with location tracking enabled.

I did a quick comparison against my own experimental code, and you see it here. Very small diff between 2.x and 1.x

![kl](https://user-images.githubusercontent.com/31855118/53771528-df08ad00-3f1d-11e9-80fd-cc4529dc3696.png)

 But what is the progress on this for you? When will 2.0 be out the door?

**Comment by KFlash** [comment]
@fkleuver Any progress on the 2.0 release?  I noticed things escalated...    So could you do the separation?  Don't bother too much with the main code except extraction.  I plan to release to public my private reference parser in the future.  Probably for v. 2.x or 3.0.
When I do so I will replace the main Cherie parser code and set a new parser standard.  Still work to fix on it to make it public friendly, but it pass and fail 110% after the spec. 

But before that we need to release 2.0.

Your thoughts?

**Comment by fkleuver** [comment]
@KFlash I'm not entirely following. Also, apologies, I've been swamped with work recently and also have some catching up to do for Aurelia as well. Too many things happening at the same time.
As for 2.0, I think we've got a dilemma here don't we? Should we do this before or after the separation of scoping?

**Comment by KFlash** [comment]
@fkleuver It's okay, and I understand your situation. Dilemma for 2.0? I'm not sure. It's only to rip out the scoping code and do what we talked about. Later when I find time for it I will *only* replace the part of Cherow that creates the AST nodes, and your scoping code will be used together with the new code. I can't see any huge issues with that? 
In my reference parser I have no scoping code at all, and no plans to add it because of the plans we talked about before.
So just focus on separation of the scoping. My reference parser code is far from finished for open source. For that we talk **in the future**. I'm still playing around with it :)

**Comment by fkleuver** [comment]
@KFlash You already had some code that took out the scoping right? Mind sharing that for reference?
Btw, if you're working on some new experimental stuff, seriously, try wasm. You've good roots in C/C++ right? You can write the parser entirely in that and compile it straight to WASM. Every major browser supports it already. It's several times faster.

**Comment by KFlash** [comment]
@fkleuver I don't have any code without scoping except for v. 1.x, but it should be easy to remove the scoping code. Just remove all code that either start or end with `scope` and also have a look at `common.ts` to see the functions related to scoping and remove them too :) 

**Note:** Tests will break ofc. - tests related to scoping. But also be aware that func params, throw statement, switch statement etc is somehow related to scoping. Easy to track it :)

WASM? I consider it after I find and end on my experimental code. Or actually it's not a experiment anymore - it's a parser equal to Cherow 2.x. But me and V8 isn't friends atm. so I tend to run into infinity loops with the code. And I'm too lazy to fix it, but it happen because I do lazy parsing, and parsing everything in reverse and call the functions I need when I need them :) I save memory on that, plus it's around 25% faster than what Cherow 1.x is. Little hard to understand the concept because it hasn't been done before in any programming language. 

But yes. I plan to convert to WASM in the end.

**Comment by KFlash** [comment]
@fkleuver Sent you a file on gitter. You can use it as a reference maybe? No export and import, but everything else should work.  Use `kleuver` instead of `cherow` and you are good to go :) Then you can test out your scoping

**Comment by KFlash** [comment]
@fkleuver How can we best simplify something like this?

```js
destructible |=
          (destructible |
            ((state.assignable | AssignmentState.NotAssignable | AssignmentState.Assignable) ^
              (AssignmentState.NotAssignable | AssignmentState.Assignable)) |
            ((AssignmentState.Assignable | AssignmentState.NotAssignable) ^ AssignmentState.Assignable) |
            DestructuringState.NotDestructible |
            DestructuringState.Assignable) ^
          (DestructuringState.NotDestructible | DestructuringState.Assignable);
        state.assignable = AssignmentState.NotAssignable;
```

**Comment by KFlash** [comment]
@fkleuver Your thoughts? I got my experimental code working now. My code is absolutely **not** optimized or tweaked, so the perf you see in the benchmark is not the finale result **But** the experimental code does much more validations than Cherow 2.x.

![newbench](https://user-images.githubusercontent.com/31855118/54176237-097fda80-44c9-11e9-929a-04323f54c245.png)

**Comment by KFlash** [comment]
@fkleuver Any timeframe for v. 2.0 release on NPM? Some people want to start using Cherow, and better to let them use v. 2.0.... See gitter for a reference.

**Comment by fkleuver** [comment]
It is already possible to try it out by installing `cherow@dev` for the record. The nightly is publishing directly from master.

**Comment by KFlash** [comment]
Maybe you should mention it on gitter? And implement jsx again?

**Comment by fkleuver** [comment]
> Maybe you should mention it on gitter?

I did.

> And implement jsx again?

Woah. Was it implemented in 1.6?

**Comment by KFlash** [comment]
JSX have always been in Cherow except 2.0. You don't remember that you tried to fix the TS types for JSX before?

**Comment by KFlash** [comment]
@fkleuver Any progress on the v. 2.0 release? See gitter

**Comment by KFlash** [comment]
@fkleuver Any release data for Cherow 2.x yet ?

**Comment by fkleuver** [comment]
@KFlash Let's clear something up here. We agreed that I would do performance optimizations, not that I would take over the project and fix your bugs as well. If there were no bugs, it could be released already because performance is just performance. Being slower than 1.6x doesn't block anyone.
If in your opinion it can't be released yet, then what's wrong?

**Comment by KFlash** [comment]
What's was holding back was the fact you talked about rip out the scoping.  If not.  Go ahead and push it to mom.  A minor release can fix performance. 

You also mentioned decorators. I guess that can be minor patch as well.

**Comment by KFlash** [comment]
Regarding the REPL you started to work on in may & june 2018. Any progress on it? Do you think it will be useful together with a 2.0 release?

**Comment by KFlash** [comment]
@fkleuver @g-plane Can you publish master to NPM? I can't access the NPM account. And close this ticket.

**Comment by g-plane** [comment]
I can't. Maybe @fkleuver can.

**Comment by KFlash** [comment]
@g-plane Maybe he can. You forgot the password to?

**Comment by g-plane** [comment]
I'm using another operating system and the Node.js haven't been installed.

**Comment by KFlash** [comment]
@fkleuver Any progress on pushing Cherow to NPM?
